app/agent/runtime/qemu/qemu.service.test.ts

```diff
@@ -0,0 +1,387 @@
+    [1, 128, name, remoteTag],
+    [1, 256, name, remoteTag],
+    [2, 128, name, remoteTag],
+    [2, 256, name, remoteTag],
+  ]),
+)(
+  `Creates vm with %d cores and %d MiB ram on %s`,
+  async (vcpuCount, memSizeMib, _name, remoteTag) =>
+    provideBlockRegistry(async ({ injector, runId, brService }) => {
+      const osIm = await brService.loadImageFromRemoteRepo(remoteTag, "osIm");
+      const osCt = await brService.createContainer(osIm, "osCt");
+      const osB = await brService.expose(osCt, EXPOSE_METHOD.BLOCK_DEV);
+      const instance = injector.resolve(Token.QemuService)(runId);
+      await instance.withRuntime(
+        {
+          ssh: {
+            username: "root",
+            password: "root",
+          },
+          vcpuCount,
+          memSizeMib,
+          fs: { "/": osB },
+          cleanupAfterStop: true,
+          shouldPoweroff: true,
+        },
+        async ({ ssh }) => {
+          const cpuInfo = (await execSshCmd({ ssh }, ["cat", "/proc/cpuinfo"])).stdout;
+          expect(cpuInfo).toMatch(new RegExp(`^cpu cores.*?${vcpuCount}$`, "gm"));
+          // Ram is more tricky and containers need a different test here
+          // Assert that the memory block size on the guest is 128MB
+          // block_size_bytes is in hex notation
+          await expect(
+            execSshCmd({ ssh }, ["cat", "/sys/devices/system/memory/block_size_bytes"]),
+          ).resolves.toHaveProperty("stdout", "8000000");
+          // Now check the amount of memory blocks on the system
+          expect(
+            (await execSshCmd({ ssh }, ["ls", "/sys/devices/system/memory/"])).stdout
+              .split("\n")
+              .filter((m) => m.startsWith("memory")),
+          ).toHaveLength(memSizeMib / 128);
+        },
+      );
+    })(),
+);
+
+test.concurrent(
+  `Doesn't hang forever if qemu fails to start`,
+  provideBlockRegistry(async ({ injector, runId, brService }) => {
+    const osCt = await brService.createContainer(void 0, "emptyCt", { mkfs: true, sizeInGB: 64 });
+    const osB = await brService.expose(osCt, EXPOSE_METHOD.BLOCK_DEV);
+    const instance = injector.resolve(Token.QemuService)(runId);
+    await expect(
+      instance.withRuntime(
+        {
+          ssh: {
+            username: "root",
+            password: "root",
+          },
+          vcpuCount: 1,
+          memSizeMib: 128,
+          fs: { "/": osB },
+          cleanupAfterStop: true,
+          shouldPoweroff: true,
+          kernelPath: "/tmp/very-legit-kernel-image",
+        },
+        async () => {},
+      ),
+    ).rejects.toThrow("closed: 1");
+  }),
+);
+
+test.concurrent(
+  `Doesn't hang forever if booting vm without /init`,
+  provideBlockRegistry(async ({ injector, runId, brService }) => {
+    const osCt = await brService.createContainer(void 0, "emptyCt", { mkfs: true, sizeInGB: 64 });
+    const osB = await brService.expose(osCt, EXPOSE_METHOD.BLOCK_DEV);
```

Comment:
hide?

Reply:
It's handled globally: https://github.com/hocus-dev/hocus/pull/93/files#diff-6728b317769e3b07f2f0926e1b27c3dadcd3958b1eb186a33b3cc8b3234f6047R162